File: p1102_fw_uploader.swift
import Foundation
import IOKit
import IOKit.usb

// Configuration Constants
let PRINTER_VID = 0x03F0
let PRINTER_PID = 0x002A
let FIRMWARE_PATH = "/Library/Printers/foo2zjs-str4ngemd/firmware/sihpP1102.dl"

func log(_ message: String) {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
    let timestamp = formatter.string(from: Date())
    print("[\(timestamp)] \(message)")
    fflush(stdout)
}

// Check if the printer has the firmware already loaded.
// When firmware is NOT loaded (bootloader mode), only 1 interface is exported (Interface 0).
// When firmware IS loaded, 2 interfaces are exported (Interface 0: Printer, Interface 1: HP EWS).
func isFirmwareLoaded(device: io_service_t) -> Bool {
    var childIterator: io_iterator_t = 0
    let kr = IORegistryEntryGetChildIterator(device, kIOServicePlane, &childIterator)
    guard kr == KERN_SUCCESS else {
        log("Warning: Failed to read device registry entry children (error \(kr))")
        return false
    }
    
    defer {
        IOObjectRelease(childIterator)
    }
    
    var interfaceCount = 0
    var child = IOIteratorNext(childIterator)
    while child != 0 {
        var className = [CChar](repeating: 0, count: 128)
        IOObjectGetClass(child, &className)
        let nameStr = String(cString: className)
        if nameStr == "IOUSBHostInterface" {
            interfaceCount += 1
        }
        IOObjectRelease(child)
        child = IOIteratorNext(childIterator)
    }
    
    return interfaceCount > 1
}

// Queries the CUPS usb backend to discover the matching printer's device URI.
func detectPrinterURI() -> String? {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/libexec/cups/backend/usb")
    let pipe = Pipe()
    process.standardOutput = pipe
    
    do {
        try process.run()
        process.waitUntilExit()
        
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        if let output = String(data: data, encoding: .utf8) {
            for line in output.components(separatedBy: .newlines) {
                if line.contains("HP") && line.contains("LaserJet") && line.contains("P1102") {
                    let parts = line.components(separatedBy: .whitespaces)
                    for part in parts {
                        if part.hasPrefix("usb://") {
                            return part
                        }
                    }
                }
            }
        }
    } catch {
        log("Error running CUPS backend to query URI: \(error)")
    }
    return nil
}

// Performs the actual firmware upload by passing the firmware file to CUPS backend via stdin.
func uploadFirmware(uri: String) -> Bool {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/libexec/cups/backend/usb")
    process.arguments = ["1", "uploader", "Upload firmware", "1", ""]
    
    var env = ProcessInfo.processInfo.environment
    env["DEVICE_URI"] = uri
    process.environment = env
    
    guard FileManager.default.fileExists(atPath: FIRMWARE_PATH) else {
        log("Error: Firmware file not found at \(FIRMWARE_PATH)")
        return false
    }
    
    guard let fileHandle = FileHandle(forReadingAtPath: FIRMWARE_PATH) else {
        log("Error: Cannot read firmware file at \(FIRMWARE_PATH)")
        return false
    }
    
    process.standardInput = fileHandle
    
    let pipeErr = Pipe()
    process.standardError = pipeErr
    
    do {
        try process.run()
        process.waitUntilExit()
        
        if process.terminationStatus == 0 {
            return true
        } else {
            let dataErr = pipeErr.fileHandleForReading.readDataToEndOfFile()
            if let errStr = String(data: dataErr, encoding: .utf8) {
                log("CUPS backend exited with status \(process.terminationStatus): \(errStr.trimmingCharacters(in: .whitespacesAndNewlines))")
            }
            return false
        }
    } catch {
        log("Failed to run CUPS backend: \(error)")
        return false
    }
}

// Handles individual printer connection event.
func handlePrinterConnected(device: io_service_t) {
    // Get Serial Number from properties for logging
    var serialNumber: String = "unknown"
    var props: Unmanaged<CFMutableDictionary>? = nil
    if IORegistryEntryCreateCFProperties(device, &props, kCFAllocatorDefault, 0) == KERN_SUCCESS,
       let dict = props?.takeRetainedValue() as? [String: Any] {
        if let serial = dict[kUSBSerialNumberString] as? String {
            serialNumber = serial
        }
    }
    
    log("Printer connection detected on USB bus (Serial: \(serialNumber)).")
    
    if isFirmwareLoaded(device: device) {
        log("Firmware is already active on this device. Skipping upload.")
        return
    }
    
    log("Firmware not active (device in bootloader mode). Commencing upload...")
    
    // Give macOS 2 seconds to establish the printer backend endpoint
    Thread.sleep(forTimeInterval: 2.0)
    
    guard let uri = detectPrinterURI() else {
        log("Error: Printer detected but failed to resolve CUPS URI. Make sure the printer queue is configured.")
        return
    }
    
    log("Resolved printer URI: \(uri)")
    log("Uploading firmware to device...")
    
    if uploadFirmware(uri: uri) {
        log("Firmware upload successful. Printer should reboot.")
    } else {
        log("Error: Firmware upload failed.")
    }
}

// Callback invoked by IOKit when a matching USB device is connected or matched.
func deviceAddedCallback(refCon: UnsafeMutableRawPointer?, iterator: io_iterator_t) {
    var device = IOIteratorNext(iterator)
    while device != 0 {
        handlePrinterConnected(device: device)
        IOObjectRelease(device)
        device = IOIteratorNext(iterator)
    }
}

// Background thread function to tail /var/log/cups/error_log for diagnostics.
func startLogMonitor() {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/tail")
    process.arguments = ["-f", "-n", "0", "/var/log/cups/error_log"]
    
    let pipe = Pipe()
    process.standardOutput = pipe
    
    do {
        try process.run()
        
        let fileHandle = pipe.fileHandleForReading
        fileHandle.readabilityHandler = { handle in
            let data = handle.availableData
            if data.isEmpty { return }
            if let output = String(data: data, encoding: .utf8) {
                for line in output.components(separatedBy: .newlines) {
                    let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
                    if trimmed.isEmpty { continue }
                    if trimmed.contains("Job ") || trimmed.contains("rastertozjs") {
                        print("[CUPS] \(trimmed)")
                        fflush(stdout)
                    }
                }
            }
        }
    } catch {
        log("Warning: Failed to start CUPS log monitor: \(error)")
    }
}

// Main Execution Entrypoint
func main() {
    log("Starting HP LaserJet P1102 Native Uploader & Monitor Daemon...")
    log("Using firmware file: \(FIRMWARE_PATH)")
    
    // Start unified diagnostics stream
    log("Monitoring CUPS error log at /var/log/cups/error_log for printer logs...")
    startLogMonitor()
    
    // Create the IOKit notification port
    let notifyPort = IONotificationPortCreate(kIOMainPortDefault)
    guard notifyPort != nil else {
        log("Fatal Error: Failed to create IOKit notification port.")
        exit(1)
    }
    
    let runLoopSource = IONotificationPortGetRunLoopSource(notifyPort).takeUnretainedValue()
    CFRunLoopAddSource(CFRunLoopGetCurrent(), runLoopSource, CFRunLoopMode.defaultMode)
    
    // Setup matching dictionary
    let matchingDict = IOServiceMatching(kIOUSBDeviceClassName) as NSMutableDictionary
    matchingDict[kUSBVendorID] = PRINTER_VID
    matchingDict[kUSBProductID] = PRINTER_PID
    
    var iterator: io_iterator_t = 0
    let kr = IOServiceAddMatchingNotification(
        notifyPort,
        kIOMatchedNotification,
        matchingDict,
        deviceAddedCallback,
        nil,
        &iterator
    )
    
    if kr != KERN_SUCCESS {
        log("Fatal Error: Failed to register IOKit service notifications (error \(kr)).")
        exit(1)
    }
    
    // Process any printer already connected before the daemon was started
    deviceAddedCallback(refCon: nil, iterator: iterator)
    
    // Run loop to keep processing notifications asynchronously
    log("Uploader daemon is active. Listening for printer USB hotplug events...")
    CFRunLoopRun()
}

main()

File: p1102_fw_uploader.py
#!/usr/bin/env python3
# Save as: print lab 2/p1102_fw_uploader.py
# Description: Simple, lightweight background daemon that monitors USB connection
#              events for the HP LaserJet Pro P1102, automatically uploads firmware,
#              and dynamically consolidates CUPS print logs in real-time.
#              Requires `pyusb` installed.

import os
import sys
import time
import subprocess
import threading
import re
import usb.core
import usb.util

PRINTER_VID = 0x03F0
PRINTER_PID = 0x002A

# Try to find the firmware file in standard installation locations
FIRMWARE_LOCATIONS = [
    "/Library/Printers/foo2zjs/firmware/sihpP1102.dl",
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "sihpP1102.dl"),
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "../sihpP1102.dl"),
    "/Users/sorce/code/hplipmaybe/sihpP1102.dl"
]

def log(msg):
    print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] {msg}", flush=True)

def find_firmware():
    for loc in FIRMWARE_LOCATIONS:
        if os.path.exists(loc):
            return loc
    return None

def detect_printer_uri():
    try:
        # Run the cups USB backend to discover the printer's connection URI
        res = subprocess.run(["/usr/libexec/cups/backend/usb"], capture_output=True, text=True)
        if res.returncode == 0:
            for line in res.stdout.splitlines():
                if "usb://" in line and ("P1102" in line or "ZJS" in line or "002A" in line.upper()):
                    parts = line.split()
                    if len(parts) >= 2:
                        uri = parts[1]
                        return uri.strip('"\'')
    except Exception as e:
        log(f"Error auto-detecting printer USB URI: {e}")
    return None

def upload_firmware(uri, fw_path):
    log(f"Uploading firmware '{fw_path}' to device URI '{uri}'...")
    env = os.environ.copy()
    env["DEVICE_URI"] = uri
    try:
        # Send raw firmware via CUPS native USB backend
        res = subprocess.run(["/usr/libexec/cups/backend/usb", "1", "", "", "1", "", fw_path], env=env, capture_output=True, text=True)
        if res.returncode == 0:
            log("Firmware upload successful. Printer should boot up.")
            return True
        else:
            log(f"Firmware upload failed with code {res.returncode}. Stderr: {res.stderr}")
    except Exception as e:
        log(f"Exception during firmware upload: {e}")
    return False

def cups_log_monitor_thread():
    log_path = "/var/log/cups/error_log"
    if not os.path.exists(log_path):
        log(f"Warning: CUPS error log not found at {log_path}. CUPS log tailing disabled.")
        return

    log(f"Monitoring CUPS error log at {log_path} for P1102 print jobs...")
    
    # Set to keep track of job IDs related to our printer
    active_jobs = set()
    
    try:
        with open(log_path, "r", errors="ignore") as f:
            # Seek to the end of the file so we only tail new entries
            f.seek(0, os.SEEK_END)
            while True:
                line = f.readline()
                if not line:
                    time.sleep(0.1)
                    continue
                
                # Check for Job ID in brackets like [Job 50]
                job_match = re.search(r"\[Job (\d+)\]", line)
                if job_match:
                    job_id = job_match.group(1)
                    
                    # Dynamically track print jobs relating to our printer queue or filter
                    if "HP_LaserJet_Professional_P1102" in line or "rastertozjs" in line:
                        active_jobs.add(job_id)
                        
                    # If this is a line for an active P1102 job, print it!
                    if job_id in active_jobs:
                        clean_line = line.strip()
                        # Strip timestamp brackets from CUPS to keep logs clean
                        clean_line = re.sub(r"^[DIEF] \[\d{2}/\w+/\d{4}:\d{2}:\d{2}:\d{2} [+-]\d{4}\] ", "", clean_line)
                        print(f"[CUPS] {clean_line}", flush=True)
                        
                    # Stop tracking job when it is completed or unloaded
                    if "Job completed" in line or "Unloading..." in line:
                        active_jobs.discard(job_id)
                        
    except Exception as e:
        log(f"Error in CUPS log monitor: {e}")

def main():
    log("Starting HP LaserJet P1102 USB Uploader & Monitor Daemon...")
    
    fw_path = find_firmware()
    if not fw_path:
        log("ERROR: Firmware file 'sihpP1102.dl' not found in any standard location:")
        for loc in FIRMWARE_LOCATIONS:
            log(f"  - {loc}")
        log("Please place the firmware file in one of these paths.")
        sys.exit(1)
        
    log(f"Using firmware: {fw_path}")
    
    # Start the CUPS log monitor thread
    log_thread = threading.Thread(target=cups_log_monitor_thread, daemon=True)
    log_thread.start()
    
    was_connected = False
    
    # Run initial check in case printer is already plugged in at startup
    try:
        dev = usb.core.find(idVendor=PRINTER_VID, idProduct=PRINTER_PID)
        if dev:
            uri = detect_printer_uri()
            if uri:
                upload_firmware(uri, fw_path)
                was_connected = True
    except Exception as e:
        log(f"Initial setup check failed: {e}")

    # Core uploader loop
    while True:
        try:
            dev = usb.core.find(idVendor=PRINTER_VID, idProduct=PRINTER_PID)
            if dev and not was_connected:
                log("Printer connection detected on USB bus.")
                # Give it a moment to initialize the descriptor tables
                time.sleep(2)
                uri = detect_printer_uri()
                if uri:
                    upload_firmware(uri, fw_path)
                    was_connected = True
                else:
                    log("Warning: Printer connected but CUPS USB backend did not return a URI yet. Retrying...")
            elif not dev and was_connected:
                log("Printer disconnected from USB bus.")
                was_connected = False
        except Exception as e:
            log(f"Error in monitor loop: {e}")
            time.sleep(5)
            continue
            
        time.sleep(2)

if __name__ == "__main__":
    main()
